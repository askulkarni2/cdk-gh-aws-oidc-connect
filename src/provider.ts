import * as iam from '@aws-cdk/aws-iam';
import { Aws, Construct } from '@aws-cdk/core';

/**
 * Represents a GitHub OIDC provider.
 */
export interface IGitHubActionsOidcProvider {
  /**
   * The ARN of the OIDC provider.
   */
  readonly providerArn: string;
}

/**
 * Defines an OIDC provider for GitHub workflows.
 *
 * Please note that only a single instance of this provider can be installed in
 * an AWS account.
 */
export class GitHubActionsOidcProvider extends Construct implements IGitHubActionsOidcProvider {
  /**
   * The OIDC domain for GitHub.
   */
  public static readonly DOMAIN = 'token.actions.githubusercontent.com';

  /**
   * The OIDC domain thumbprint for GitHub.
   */
  public static readonly THUMBPRINT = 'a031c46782e6e6c662c2c87c76da9aa62ccabd8e';

  /**
   * @param account The AWS account for which you want to obtain the OIDC
   * provider. If not specified, we will use the current account.
   *
   * @returns The singleton GitHub OIDC provider for an account.
   */
  public static forAccount(account?: string): IGitHubActionsOidcProvider {
    return {
      providerArn: arnForAccount(account),
    };
  }

  /**
   * The ARN of the OIDC provider.
   */
  public readonly providerArn: string;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    const provider = new iam.CfnOIDCProvider(scope, `${id}.GithubOidcProvider`, {
      url: `https://${GitHubActionsOidcProvider.DOMAIN}`,
      thumbprintList: [GitHubActionsOidcProvider.THUMBPRINT],
      clientIdList: ['sigstore'],
    });

    this.providerArn = provider.attrArn;
  }
}

function arnForAccount(account?: string): string {
  account = account ?? Aws.ACCOUNT_ID;
  return `arn:aws:iam::${account}:oidc-provider/${GitHubActionsOidcProvider.DOMAIN}`;
}