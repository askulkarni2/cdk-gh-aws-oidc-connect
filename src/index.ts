import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import { GitHubActionsOidcProvider } from './provider';
import { GitHubActionsRole } from './role';

export * from './provider';
export * from './role';

export interface GitHubActionsAwsOidcConnectProps {
  /**
   * GitHub repository which will assume the IAM role.
   * E.g. aidansteele/aws-federation-github-actions
   */
  readonly repo: string;

  /**
   * The name of the IAM role to create.
   *
   * @default - generated
   */
  readonly roleName?: string;

  /**
   * A list of IAM policies.
   */
  readonly policies?: Array<iam.Policy>;

  /**
   * A list of Managed IAM policies.
   */
  readonly managedPolicies?: Array<iam.IManagedPolicy>;

  /**
   * External IDs required in order to assume the role.
   * @default - not limited to specific external IDs.
   */
  readonly externalIds?: string[];
}

export class GitHubActionsAwsOidcConnect extends cdk.Construct {
  /**
   * This is the role GitHub Actions can assume for the repos provided.
   */
  public roleToAssume: iam.Role;

  constructor(scope: cdk.Construct, id: string, props: GitHubActionsAwsOidcConnectProps) {
    super(scope, id);

    const provider = new GitHubActionsOidcProvider(this, 'Provider');

    this.roleToAssume = new GitHubActionsRole(this, 'iam-role-to-assume', {
      provider: provider,
      managedPolicies: props.managedPolicies,
      repository: props.repo,
      roleName: props.roleName,
      externalIds: props.externalIds,
    });

    // Attach any other inline policies provided
    if (props.policies && props.policies.length >0 ) {
      props.policies.forEach((policy) => {
        this.roleToAssume.attachInlinePolicy(policy);
      });
    }
  }
}