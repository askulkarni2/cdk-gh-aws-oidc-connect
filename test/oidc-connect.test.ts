import { SynthUtils } from '@aws-cdk/assert';
import * as ecr from '@aws-cdk/aws-ecr';
import * as iam from '@aws-cdk/aws-iam';
import { Stack } from '@aws-cdk/core';
import * as oidcConnect from '../src/index';

test('oidcConnect creates a role', () => {
  const stack = new Stack();
  new oidcConnect.GitHubActionsAwsOidcConnect(stack, 'oidc-connect', {
    repos: ['askulkarni2/cdk-gh-aws-oidc-connect'],
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('oidcConnect creates a role with a managed policy', () => {
  const stack = new Stack();
  new oidcConnect.GitHubActionsAwsOidcConnect(stack, 'oidc-connect', {
    repos: ['askulkarni2/cdk-gh-aws-oidc-connect'],
    managedPolicies: [
      iam.ManagedPolicy.fromAwsManagedPolicyName('ReadOnlyAccess'),
    ],
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('oidcConnect creates a role with an inline policy', () => {
  const stack = new Stack();
  new oidcConnect.GitHubActionsAwsOidcConnect(stack, 'oidc-connect', {
    repos: ['askulkarni2/cdk-gh-aws-oidc-connect'],
    policies: [
      new iam.Policy(stack, 'inline-policy'),
    ],
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

test('oidcConnect role is granted access to an ECR repo', () => {
  const stack = new Stack();
  const oidcProvider = new oidcConnect.GitHubActionsAwsOidcConnect(stack, 'oidc-connect', {
    repos: ['askulkarni2/cdk-gh-aws-oidc-connect'],
  });
  const ecrRepo = new ecr.Repository(stack, 'test-repo');
  ecrRepo.grantPullPush(oidcProvider.roleToAssume);
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});