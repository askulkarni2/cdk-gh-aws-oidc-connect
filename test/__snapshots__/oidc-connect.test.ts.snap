// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oidcConnect creates a role 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15ArtifactHashD83177FF": Object {
      "Description": "Artifact hash for asset \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3Bucket2010B486": Object {
      "Description": "S3 bucket for asset \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD": Object {
      "Description": "S3 key for asset version \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": Object {
      "DependsOn": Array [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3Bucket2010B486",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "githuboidcprovider510B14A8": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ClientIDList": Array [
          "sigstore",
        ],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://vstoken.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "iamroletoassume09F64513": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringEquals": Object {
                  "vstoken.actions.githubusercontent.com:sub": Array [
                    "repo:askulkarni2/cdk-gh-aws-oidc-connect:*",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": Object {
                  "Ref": "githuboidcprovider510B14A8",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`oidcConnect creates a role with a managed policy 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15ArtifactHashD83177FF": Object {
      "Description": "Artifact hash for asset \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3Bucket2010B486": Object {
      "Description": "S3 bucket for asset \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD": Object {
      "Description": "S3 key for asset version \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": Object {
      "DependsOn": Array [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3Bucket2010B486",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "githuboidcprovider510B14A8": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ClientIDList": Array [
          "sigstore",
        ],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://vstoken.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "iamroletoassume09F64513": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringEquals": Object {
                  "vstoken.actions.githubusercontent.com:sub": Array [
                    "repo:askulkarni2/cdk-gh-aws-oidc-connect:*",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": Object {
                  "Ref": "githuboidcprovider510B14A8",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/ReadOnlyAccess",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`oidcConnect creates a role with an inline policy 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15ArtifactHashD83177FF": Object {
      "Description": "Artifact hash for asset \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3Bucket2010B486": Object {
      "Description": "S3 bucket for asset \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD": Object {
      "Description": "S3 key for asset version \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": Object {
      "DependsOn": Array [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3Bucket2010B486",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "githuboidcprovider510B14A8": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ClientIDList": Array [
          "sigstore",
        ],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://vstoken.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "iamroletoassume09F64513": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringEquals": Object {
                  "vstoken.actions.githubusercontent.com:sub": Array [
                    "repo:askulkarni2/cdk-gh-aws-oidc-connect:*",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": Object {
                  "Ref": "githuboidcprovider510B14A8",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`oidcConnect role is granted access to an ECR repo 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15ArtifactHashD83177FF": Object {
      "Description": "Artifact hash for asset \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3Bucket2010B486": Object {
      "Description": "S3 bucket for asset \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
    "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD": Object {
      "Description": "S3 key for asset version \\"ff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0": Object {
      "DependsOn": Array [
        "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3Bucket2010B486",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersff8d00f9cedcafaf1ab299446ed2500b61b0b88ab40f066d05b26ff66bd58b15S3VersionKey9B3E73DD",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderRole517FED65": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iam:CreateOpenIDConnectProvider",
                    "iam:DeleteOpenIDConnectProvider",
                    "iam:UpdateOpenIDConnectProviderThumbprint",
                    "iam:AddClientIDToOpenIDConnectProvider",
                    "iam:RemoveClientIDFromOpenIDConnectProvider",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "githuboidcprovider510B14A8": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "ClientIDList": Array [
          "sigstore",
        ],
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomAWSCDKOpenIdConnectProviderCustomResourceProviderHandlerF2C543E0",
            "Arn",
          ],
        },
        "Url": "https://vstoken.actions.githubusercontent.com",
      },
      "Type": "Custom::AWSCDKOpenIdConnectProvider",
      "UpdateReplacePolicy": "Delete",
    },
    "iamroletoassume09F64513": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringEquals": Object {
                  "vstoken.actions.githubusercontent.com:sub": Array [
                    "repo:askulkarni2/cdk-gh-aws-oidc-connect:*",
                  ],
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": Object {
                  "Ref": "githuboidcprovider510B14A8",
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "iamroletoassumeDefaultPolicyA1C20718": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "ecr:BatchCheckLayerAvailability",
                "ecr:GetDownloadUrlForLayer",
                "ecr:BatchGetImage",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testrepo76ABFB50",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "ecr:GetAuthorizationToken",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "ecr:PutImage",
                "ecr:InitiateLayerUpload",
                "ecr:UploadLayerPart",
                "ecr:CompleteLayerUpload",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "testrepo76ABFB50",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "iamroletoassumeDefaultPolicyA1C20718",
        "Roles": Array [
          Object {
            "Ref": "iamroletoassume09F64513",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testrepo76ABFB50": Object {
      "DeletionPolicy": "Retain",
      "Type": "AWS::ECR::Repository",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
